R/regression.R

```diff
@@ -1617,14 +1634,38 @@ removeMissingVariables <- function(data, formula, formula.with.interaction,
 # Processes the unstacked data and stacks it.
 # If the stacking is successful, the interaction, subset, weights are also updated
 # to be the appropriate size
-processAndStackData <- function(unstacked.data, formula, interaction, subset, weights)
+processAndStackData <- function(unstacked.data, formula, interaction, subset, weights, missing)
 {
     checkDataFeasibleForStacking(unstacked.data)
     unstacked.data <- removeDataReduction(unstacked.data)
     validated.unstacked.output <- validateDataForStacking(unstacked.data)
     unstacked.data <- validated.unstacked.output[["data"]]
     stacks <- validated.unstacked.output[["stacks"]]
-    data <- stackData(unstacked.data)
+    stacked.data <- stackData(unstacked.data)
+    n.orig.stacked.cases = nrow(stacked.data)
+
+    # Exclude missing cases. Stacking can result in
+    # many empty cases when data set is sparse. Respondent-level
+    # functions (e.g. predicted values) are not supported
+    # for stacked data, so no need to keep track of removed
+    # cases.
+    rm.missing <- rep(FALSE, n.orig.stacked.cases)
+    # Don't exclude missing cases when using pairwise correlations.
+    # It currently causes differences which need to be investigated further. (DS-4844)
+    if (missing != "Use partial data (pairwise correlations") {
+    	missing.vals <- lapply(stacked.data, is.na)
+	    .reduceFunction <- if (missing == "Exclude cases with missing data") `|` else `&`
+	    rm.missing <- Reduce(.reduceFunction, missing.vals)	
+    }
+    
+
+    if (all(rm.missing)) {
+        stop("The stacked data contains no observations after missing data has been removed.")
+    }
+
+    data <- stacked.data[!rm.missing, ]
+    data <- CopyAttributes(data, stacked.data)
```

Comment:
Could put this in a `if` statement that checks `any(rm.missing)` to prevent subscripting if not required. On that note, the calculation `any(rm.missing)` is used many times throughout and could be reused if defined with another variable.